### PR TITLE
SimpleCAST Verifier

### PR DESCRIPTION
**Pull request by kriw**
Related to #156 

- Changed `ValueEdge::GotoDst` to `ActionEdge::GotoDst`
- Added `verify` to tests

### CODE REVIEW

**Comment by sushant94** [comment]
Is this ready to be merged? SGTM so far.

**Comment by kriw** [comment]
Yes, this is ready to be merged.

**Comment by sushant94** [comment]
@HMPerson1 I think this is a small enough PR to start from. Please review this :)

File: src/backend/lang_c/c_simple_ast.rs [hunk 1]
```diff
@@ -525,8 +525,149 @@ impl SimpleCAST {
...
+    // 2. There are ActionEdge::IfThen, ValueEdge::Conditional from ActionNode::If
+    fn verify_if(node: NodeIndex, cast: &SimpleCAST) -> Result<(), String> {
+        match cast.ast.node_weight(node) {
+            Some(&SimpleCASTNode::Action(ActionNode::If)) => {},
+            _ => return Ok(()),
+        };
+        let mut ret = String::new();
+        let cond = cast.branch_condition(node);
+        if cond.is_none() {
+            ret = format!("{}; No condition node is found.", ret);
```
Comment (HMPerson1) [inline src/backend/lang_c/c_simple_ast.rs]: `ret` is empty before here, so after this statement `ret` will start with `"; ..."`. I don't think that's what you intended.

File: src/backend/lang_c/c_simple_ast.rs [hunk 2]
```diff
@@ -525,8 +525,149 @@ impl SimpleCAST {
...
+        let edges = cast.ast
+            .edges_directed(node, Direction::Outgoing).collect::<Vec<_>>();
+        for edge in edges {
+            match (edge.weight(), cast.ast.node_weight(edge.target())) {
+                (SimpleCASTEdge::Action(_), Some(&SimpleCASTNode::Action(_))) => {},
+                (SimpleCASTEdge::Value(_), Some(&SimpleCASTNode::Value(_))) => {},
+                (SimpleCASTEdge::Action(ActionEdge::GotoDst), Some(&SimpleCASTNode::Entry)) => {},
+                _ => {
+                    is_err = true;
+                    ret = format!("{}; {:?} {:?}", ret, edge.weight(),
```
Comment (HMPerson1) [inline src/backend/lang_c/c_simple_ast.rs]: Again, `ret` may be empty before this, in which case `ret` will start with `"; ..."`.

File: src/backend/lang_c/c_simple_ast.rs [hunk 3]
```diff
@@ -525,8 +525,149 @@ impl SimpleCAST {
...
+    }
+
+    // 1. There are at most 1 ActionEdge::Normal from each node.
+    fn verify_normal_action(node: NodeIndex, cast: &SimpleCAST) -> Result<(), String> {
+        match cast.ast.node_weight(node) {
+            Some(&SimpleCASTNode::Action(_)) => {},
+            _ => return Ok(()),
+        };
+        let normal_actions = cast.ast.edges_directed(node, Direction::Outgoing)
+            .into_iter()
```
Comment (HMPerson1) [inline src/backend/lang_c/c_simple_ast.rs]: `edges_directed(..)` already returns an `Iterator`; no need to call `into_iter()`.

File: src/backend/lang_c/c_simple_ast.rs [hunk 4]
```diff
@@ -525,8 +525,149 @@ impl SimpleCAST {
...
+    fn verify_goto(node: NodeIndex, cast: &SimpleCAST) -> Result<(), String> {
+        match cast.ast.node_weight(node) {
+            Some(&SimpleCASTNode::Action(ActionNode::Goto)) => {},
+            _ => return Ok(()),
+        };
+        let mut is_err = false;
+        let mut ret = String::new();
+        let gotos = cast.ast
+            .edges_directed(node, Direction::Outgoing)
+            .into_iter()
```
Comment (HMPerson1) [inline src/backend/lang_c/c_simple_ast.rs]: Again, unnecessary `into_iter()`.

File: src/backend/lang_c/c_simple_ast.rs [hunk 5]
```diff
@@ -728,6 +869,7 @@ mod test {
         let entry = ast.entry;
         let assn = ast.assign(x, y, entry);
         let _ = ast.call_func("func", &[z, w], assn, None);
+        SimpleCASTVerifier::verify(&ast).expect("This should be Ok");
```
Comment (HMPerson1) [inline src/backend/lang_c/c_simple_ast.rs]: Maybe something like "SimpleCAST verification failed"? Same for all the copies.

File: src/backend/lang_c/c_simple_ast.rs [hunk 6]
```diff
@@ -525,8 +525,149 @@ impl SimpleCAST {
...
+        match cast.ast.node_weight(node) {
+            Some(&SimpleCASTNode::Action(ActionNode::If)) => {},
+            _ => return Ok(()),
+        };
+        let mut ret = String::new();
+        let cond = cast.branch_condition(node);
+        if cond.is_none() {
+            ret = format!("{}; No condition node is found.", ret);
+        }
+        let branches = cast.branch(node).map(|x| x.clone());
```
Comment (HMPerson1) [inline src/backend/lang_c/c_simple_ast.rs]: Content of `Option<_>` is never observed; `clone()` is unnecessary.

File: src/backend/lang_c/c_simple_ast.rs [hunk 7]
```diff
@@ -525,8 +525,149 @@ impl SimpleCAST {
...
+            Ok(())
+        }
+    }
+
+    // 3. The targets of value edges are ValueNode, The target of action edges are ActionNode.
+    fn verify_edge_action(node: NodeIndex, cast: &SimpleCAST) -> Result<(), String> {
+        let mut is_err = false;
+        let mut ret = String::new();
+        let edges = cast.ast
+            .edges_directed(node, Direction::Outgoing).collect::<Vec<_>>();
```
Comment (HMPerson1) [inline src/backend/lang_c/c_simple_ast.rs]: `collect()` is unnecessary; you only iterate once over it

File: src/backend/lang_c/c_simple_ast.rs [hunk 8]
```diff
@@ -525,8 +525,149 @@ impl SimpleCAST {
...
+            _ => return Ok(()),
+        };
+        let normal_actions = cast.ast.edges_directed(node, Direction::Outgoing)
+            .into_iter()
+            .filter(|e| {
+                match e.weight() {
+                    &SimpleCASTEdge::Action(ActionEdge::Normal) => true,
+                    _ => false,
+                }
+            }).collect::<Vec<_>>();
```
Comment (HMPerson1) [inline src/backend/lang_c/c_simple_ast.rs]: `collect()` only to check length; use `count()` instead.

**Pull request merged**